cli/js/testing.ts

```diff
@@ -110,7 +228,56 @@ function filterTests(
     }
 
     return passes;
-  });
+  };
+}
+
+interface TestReporter {
+  start(msg: StartMsg): Promise<void>;
+  test(msg: TestMsg): Promise<void>;
+  end(msg: EndMsg): Promise<void>;
+}
+
+class ConsoleReporter implements TestReporter {
+  constructor(private console: Console) {}
+
+  async start(msg: StartMsg): Promise<void> {
+    this.console.log(`running ${msg.tests} tests`);
+  }
+
+  async test(msg: TestMsg): Promise<void> {
+    const { result } = msg;
+
+    if (result.passed) {
+      this.console.log(
+        `${GREEN_OK}     ${result.name} ${formatDuration(result.duration)}`
+      );
+    } else {
+      this.console.log(`${RED_FAILED} ${result.name}`);
+      this.console.log(result.error!.stack);
+    }
+  }
+
+  async end(msg: EndMsg): Promise<void> {
+    const stats = msg.stats;
+    // Attempting to match the output of Rust's test runner.
```

Comment:
Is this the case for `test` too? :crossed_fingers:  (not necessarily this PR but is that a goal?)

Reply:
@hayd not sure I understand. This hasn't changed, it's just been moved

Reply:
Atm the last line is similar to `cargo test`, but the per-test output is not similar. I think it should be.

There were comments here about capturing output too, so that stdout/stderr are only printed (at the end) if the tests fail. (Without interlacing the test result report.)

Reply:
> Atm the last line is similar to `cargo test`, but the per-test output is not similar. I think it should be.

Okay, now I get it, yeah that's probable, @ry is in favor of that AFAIK

> There were comments here about capturing output too, so that stdout/stderr are only printed (at the end) if the tests fail. (Without interlacing the test result report.)

We discussed that and there's not decision yet; if we freeze `Deno` global deeply there will be no way to mock `Deno.stdout`/`Deno.stderr` and there's always `Deno.core.print()` that users could use...

Anyway, this PR is just a prerequisite for #4336

Reply:
Is there an issue for the freezing?
I rely on mutability of console.log etc. in deno-lambda (to make them work with cloudwatch logs).

Reply:
No, there is none, but I'm talking about `Deno` global, `console` should be writeable like in browser